File: cortex/webgraph/contracts.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Literal

from pydantic import BaseModel, Field

WebGraphMode = Literal["semantic", "episodic", "hybrid"]


class WebGraphCapabilities(BaseModel):
    filters: bool = True
    subgraph: bool = True
    open_file: bool = True
    relation_explanations: bool = True


class WebGraphStats(BaseModel):
    node_count: int = 0
    edge_count: int = 0
    mode: WebGraphMode = "hybrid"
    truncated: bool = False


class WebGraphNode(BaseModel):
    id: str
    node_type: str
    source: Literal["semantic", "episodic"]
    label: str
    summary: str = ""
    rel_path: str | None = None
    memory_id: str | None = None
    tags: list[str] = Field(default_factory=list)
    files: list[str] = Field(default_factory=list)
    timestamp: str | None = None
    degree: int = 0
    metadata: dict[str, Any] = Field(default_factory=dict)


class WebGraphEdge(BaseModel):
    id: str
    source: str
    target: str
    edge_type: str
    weight: float = 1.0
    evidence: list[str] = Field(default_factory=list)


class WebGraphSnapshot(BaseModel):
    version: str = "2.0"
    fingerprint: str
    generated_at: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat(),
    )
    mode: WebGraphMode = "hybrid"
    stats: WebGraphStats
    capabilities: WebGraphCapabilities = Field(default_factory=WebGraphCapabilities)
    nodes: list[WebGraphNode] = Field(default_factory=list)
    edges: list[WebGraphEdge] = Field(default_factory=list)


class WebGraphNodeDetail(BaseModel):
    node: WebGraphNode
    relations: list[WebGraphEdge] = Field(default_factory=list)
    neighbors: list[WebGraphNode] = Field(default_factory=list)


class SemanticRecord(BaseModel):
    node_id: str
    node_type: str
    title: str
    summary: str
    rel_path: str
    abs_path: str
    tags: list[str] = Field(default_factory=list)
    links: list[str] = Field(default_factory=list)
    content: str = ""
    embedding: list[float] | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class EpisodicRecord(BaseModel):
    node_id: str
    node_type: str
    label: str
    summary: str
    memory_id: str
    tags: list[str] = Field(default_factory=list)
    files: list[str] = Field(default_factory=list)
    timestamp: str | None = None
    content: str = ""
    metadata: dict[str, Any] = Field(default_factory=dict)
    embedding: list[float] | None = None

File: rust/crates/cortex-webgraph-server/src/contracts.rs
//! Modelos del WebGraph — porteo de `cortex/webgraph/contracts.py`.

#![forbid(unsafe_code)]

use serde::{Deserialize, Serialize};
use serde_json::Value;

pub type WebGraphMode = String; // "semantic" | "episodic" | "hybrid"

/// Literales válidos de `WebGraphMode` (get_args en Python).
pub const WEBGRAPH_MODES: &[&str] = &["semantic", "episodic", "hybrid"];

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebGraphCapabilities {
    #[serde(default = "yes")]
    pub filters: bool,
    #[serde(default = "yes")]
    pub subgraph: bool,
    #[serde(default = "yes")]
    pub open_file: bool,
    #[serde(default = "yes")]
    pub relation_explanations: bool,
}

fn yes() -> bool {
    true
}

impl Default for WebGraphCapabilities {
    fn default() -> Self {
        Self {
            filters: true,
            subgraph: true,
            open_file: true,
            relation_explanations: true,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebGraphStats {
    #[serde(default)]
    pub node_count: i64,
    #[serde(default)]
    pub edge_count: i64,
    #[serde(default = "default_mode")]
    pub mode: String,
    #[serde(default)]
    pub truncated: bool,
}

fn default_mode() -> String {
    "hybrid".into()
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebGraphNode {
    pub id: String,
    pub node_type: String,
    /// "semantic" | "episodic"
    pub source: String,
    pub label: String,
    #[serde(default)]
    pub summary: String,
    #[serde(default)]
    pub rel_path: Option<String>,
    #[serde(default)]
    pub memory_id: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub files: Vec<String>,
    #[serde(default)]
    pub timestamp: Option<String>,
    #[serde(default)]
    pub degree: i64,
    #[serde(default)]
    pub metadata: serde_json::Map<String, Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebGraphEdge {
    pub id: String,
    pub source: String,
    pub target: String,
    pub edge_type: String,
    #[serde(default = "one")]
    pub weight: f64,
    #[serde(default)]
    pub evidence: Vec<String>,
}

fn one() -> f64 {
    1.0
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebGraphSnapshot {
    #[serde(default = "version")]
    pub version: String,
    pub fingerprint: String,
    /// ISO-8601 UTC con microsegundos (normalizado {{TS}} en el gate).
    #[serde(default)]
    pub generated_at: String,
    #[serde(default = "default_mode")]
    pub mode: String,
    pub stats: WebGraphStats,
    #[serde(default)]
    pub capabilities: WebGraphCapabilities,
    #[serde(default)]
    pub nodes: Vec<WebGraphNode>,
    #[serde(default)]
    pub edges: Vec<WebGraphEdge>,
}

fn version() -> String {
    "2.0".into()
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebGraphNodeDetail {
    pub node: WebGraphNode,
    #[serde(default)]
    pub relations: Vec<WebGraphEdge>,
    #[serde(default)]
    pub neighbors: Vec<WebGraphNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SemanticRecord {
    pub node_id: String,
    pub node_type: String,
    pub title: String,
    pub summary: String,
    pub rel_path: String,
    pub abs_path: String,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub links: Vec<String>,
    #[serde(default)]
    pub content: String,
    #[serde(default)]
    pub embedding: Option<Vec<f64>>,
    #[serde(default)]
    pub metadata: serde_json::Map<String, Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EpisodicRecord {
    pub node_id: String,
    pub node_type: String,
    pub label: String,
    pub summary: String,
    pub memory_id: String,
    #[serde(default)]
    pub tags: Vec<String>,
    #[serde(default)]
    pub files: Vec<String>,
    #[serde(default)]
    pub timestamp: Option<String>,
    #[serde(default)]
    pub content: String,
    #[serde(default)]
    pub metadata: serde_json::Map<String, Value>,
    #[serde(default)]
    pub embedding: Option<Vec<f64>>,
}
